src/main/java/org/opensearch/knn/index/query/KNNQueryXContentHelper.java

```diff
@@ -0,0 +1,205 @@
+     *  "my_vector2": {
+     *      "vector": [2, 3, 5, 6],
+     *      "k": 2,
+     *      ...
+     *   }
+     * to simplify the parsing process, we can define an object parser that will the internal structure after the
+     * field name. We cannot unfortunately also parse the field name because it ends up in the same structure
+     * as the nested portion. So we need to do that separately.
+     */
+    private static ObjectParser<KNNQueryBuilder.Builder, Void> createInternalObjectParser() {
+        ObjectParser<KNNQueryBuilder.Builder, Void> internalParser = new ObjectParser<>(NAME, KNNQueryBuilder.Builder::new);
+        internalParser.declareFloat(KNNQueryBuilder.Builder::boost, BOOST_FIELD);
+        internalParser.declareString(KNNQueryBuilder.Builder::queryName, NAME_FIELD);
+        internalParser.declareFloatArray((b, v) -> b.vector(floatListToFloatArray(v)), VECTOR_FIELD);
+        internalParser.declareInt(KNNQueryBuilder.Builder::k, K_FIELD);
+        internalParser.declareBoolean((b, v) -> {
+            if (isClusterOnOrAfterMinRequiredVersion("ignore_unmapped")) {
+                b.ignoreUnmapped(v);
+            }
+        }, IGNORE_UNMAPPED_FIELD);
+        internalParser.declareFloat(KNNQueryBuilder.Builder::maxDistance, MAX_DISTANCE_FIELD);
+        internalParser.declareFloat(KNNQueryBuilder.Builder::minScore, MIN_SCORE_FIELD);
+
+        internalParser.declareObject(
+            KNNQueryBuilder.Builder::methodParameters,
+            (p, v) -> MethodParametersParser.fromXContent(p),
+            METHOD_PARAMS_FIELD
+        );
+        internalParser.declareObject(KNNQueryBuilder.Builder::filter, (p, v) -> parseInnerQueryBuilder(p), FILTER_FIELD);
+
+        return internalParser;
+    }
+
+    /**
+     * Convert XContent to KNNQueryBuilder
+     *
+     * @param parser input parser
+     * @return KNNQueryBuilder
+     * @throws IOException on parsing failure
+     */
+    public static KNNQueryBuilder fromXContent(XContentParser parser) throws IOException {
+        String fieldName = null;
+        String currentFieldName = null;
+        XContentParser.Token token;
+        KNNQueryBuilder.Builder builder = null;
+        List<Object> vector = null;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                throwParsingExceptionOnMultipleFields(parser.getTokenLocation(), fieldName, currentFieldName);
+                fieldName = currentFieldName;
+                builder = INTERNAL_PARSER.apply(parser, null);
+            } else {
+                throwParsingExceptionOnMultipleFields(parser.getTokenLocation(), fieldName, parser.currentName());
+                fieldName = parser.currentName();
+                vector = parser.list();
```

Comment:
Is there a need to check if the `fieldName.equals(VECTOR_FIELD.getName())`

Reply:
Good catch, that makes sense to have.

Reply:
Wait, nvm. We do not do this now. I dont want to change behavior in this PR. See https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java#L409